src/chromo/decay_afterburner.py

```diff
@@ -0,0 +1,96 @@
+import chromo
+import numpy as np
+
+
+class DecayAfterburner:
+    def __init__(self, seed=None, stable_pids=None, decaying_pids=None):
+        """Initialize pythia8 and set `stable` and `decaying` particle types.
+
+        `decaying_pids` overrides `stable_pids` if the same pid is given
+        in both lists. If both are `None`, Pythia8 defaults are applied
+        """
+        config = ["ProcessLevel:all = off", "ParticleDecays:tau0Max = 1e100"]
+        pythia8 = chromo.models.Pythia8(None, seed=seed, config=config, banner=False)
+        self.pythia = pythia8._pythia
+
+        if stable_pids is not None:
+            for pid in stable_pids:
+                self.pythia.particleData.mayDecay(pid, False)
+
+        if decaying_pids is not None:
+            for pid in decaying_pids:
+                self.pythia.particleData.mayDecay(pid, True)
+
+        self.stable_pids = stable_pids
+        self.decaying_pids = decaying_pids
+
+    def __call__(self, event):
+        """Decay particles in `event` which are set as `decaying`.
+
+        Particle is `decaying` if its pdg id is passed
+        in `decaying_pids` list of constructor or it is decaing by default
+        in Pythia8. `stable_pids` of constructor is used to set stable
+        particles.
+
+        After the call `event` contains initial particles + produced particles
+        The decayed particles has event.status == 2.
+        """
+        if len(event.pid) == 0:
+            return
+
+        # Save not_final status
+        not_final = event.status != 1
+        status = np.copy(event.status[not_final])
+
+        # Particles with negative status are not processed by Pythia
+        event.status[not_final] = -1
+
+        # Save original event parameters which are not passed
+        # to Pythia8 stack
+        charge = event.charge
+        vx = event.vx
+        vy = event.vy
+        vz = event.vz
+        vt = event.vt
+
+        mothers = event.mothers if event.mothers is not None else None
+        daughters = event.daughters if event.daughters is not None else None
+        init_len = len(event.pid)
+
+        # Put on stack
```

Comment:
I'm wondering if it is safer to call event.reset() beforehand to always start with a clean stack, instead of the backup/restore.

Reply:
event.reset() is called by C++ code, so only clean stack is used. I copy the parameters that are not passed to Pythia. The first init_len particles returned by Pythia is the original particles of the event. Pythia assignes some values to vx, vy, vz, vt, which are different from original values. Statuses of decayed particles are set to 2. Here I save the statuses given by original event generator. Charges are saved, because some particles have `pid` that are unknown to Pythia and it will return some default value (777) in this case.